builtin/logical/database/rotation.go

```diff
@@ -0,0 +1,380 @@
+				// no static role exists for this wal entry, delete it
+				if err := framework.DeleteWAL(ctx, s, walEntry.walID); err != nil {
+					log.Warn("unable to delete WAL", "error", err, "WAL ID", walEntry.walID)
+				}
+			}
+			continue
+		}
+		item := queue.Item{
+			Key:      roleName,
+			Priority: role.StaticAccount.LastVaultRotation.Add(role.StaticAccount.RotationPeriod).Unix(),
+		}
+
+		// check if role name is in map
+		if walEntry != nil {
+			// check walEntry last vault time
+			if !walEntry.LastVaultRotation.IsZero() && walEntry.LastVaultRotation.Before(role.StaticAccount.LastVaultRotation) {
+				// WAL's last vault rotation record is older than the role's data, so
+				// delete and move on
+				if err := framework.DeleteWAL(ctx, s, walEntry.walID); err != nil {
+					log.Warn("unable to delete WAL", "error", err, "WAL ID", walEntry.walID)
+				}
+			} else {
+				log.Info("adjusting priority for Role")
+				item.Value = walEntry.walID
+				item.Priority = time.Now().Unix()
+			}
+		}
+
+		if err := b.pushItem(&item); err != nil {
+			log.Warn("unable to enqueue item", "error", err, "role", roleName)
+		}
+	}
+}
+
+// runTicker kicks off a periodic ticker that invoke the automatic credential
+// rotation method at a determined interval. The default interval is 5 seconds.
+func (b *databaseBackend) runTicker(ctx context.Context, s logical.Storage) {
+	b.logger.Info("starting periodic ticker")
+	tick := time.NewTicker(queueTickInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			b.rotateCredentials(ctx, s)
+
+		case <-ctx.Done():
+			b.logger.Info("stopping periodic ticker")
+			return
+		}
+	}
+}
+
+// setCredentialsWAL is used to store information in a WAL that can retry a
+// credential setting or rotation in the event of partial failure.
+type setCredentialsWAL struct {
+	NewPassword string
+	OldPassword string
+	RoleName    string
+	Username    string
+
+	LastVaultRotation time.Time
+
+	walID string
+}
+
+// rotateCredentials sets a new password for a static account. This method is
+// invoked in the runTicker method, which is in it's own go-routine, and invoked
+// periodically (approximately every 5 seconds).
+//
+// This method loops through the priority queue, popping the highest priority
+// item until it encounters the first item that does not yet need rotation,
+// based on the current time.
+func (b *databaseBackend) rotateCredentials(ctx context.Context, s logical.Storage) error {
+	for {
+		// quit rotating credentials if shutdown has started
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		item, err := b.popItem()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// guard against possible nil item
+		if item == nil {
+			return nil
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn("unable load role", "role", item.Key, "error", err)
+			continue
```

Comment:
Do we want to push the item back onto the queue here; with some sort of backoff?

Reply:
That sounds right. Added a 10 second increment to the priority in 3cf4b4d3a